From 68aa9ac337536d6286eebeafe1ce7620e3ac7a15 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Mon, 23 Jun 2014 21:49:47 -0400
Subject: [fs] nfsd: session needs room for following op to error out

Message-id: <1403560314-32525-46-git-send-email-bfields@redhat.com>
Patchwork-id: 84475
O-Subject: [Patch RHEL7 045/172] nfsd4: session needs room for following op to error out
Bugzilla: 1111719
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

Cc: stable@vger.kernel.org
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 4c69d5855a16f7378648c5733632628fa10431db
Bugzilla: 1111719
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4proc.c b/fs/nfsd/nfs4proc.c
index ef76ba6..e47c036 100644
--- a/fs/nfsd/nfs4proc.c
+++ b/fs/nfsd/nfs4proc.c
@@ -1358,6 +1358,12 @@ nfsd4_proc_compound(struct svc_rqst *rqstp,
   /* If op is non-idempotent */
   if (opdesc->op_flags & OP_MODIFIES_SOMETHING) {
    plen = opdesc->op_rsize_bop(rqstp, op);
+   /*
+    * If there's still another operation, make sure
+    * we'll have space to at least encode an error:
+    */
+   if (resp->opcnt < args->opcnt)
+    plen += COMPOUND_ERR_SLACK_SPACE;
    op->status = nfsd4_check_resp_size(resp, plen);
   }
 
-- 
1.7.1